src/node/stage/joining.rs

```diff
@@ -12,74 +12,67 @@ use crate::{
     event::Connected,
     id::P2pNode,
     messages::{
-        self, BootstrapResponse, JoinRequest, Message, MessageAction, Variant, VerifyStatus,
+        self, BootstrapResponse, JoinRequest, Message, MessageStatus, QueuedMessage, Variant,
+        VerifyStatus,
     },
     relocation::RelocatePayload,
     section::EldersInfo,
     xor_space::Prefix,
 };
-use bytes::Bytes;
-use std::time::Duration;
+use std::{mem, net::SocketAddr, time::Duration};
 
 /// Time after which an attempt to joining a section is cancelled (and possibly retried).
-pub const JOIN_TIMEOUT: Duration = Duration::from_secs(600);
+pub const JOIN_TIMEOUT: Duration = Duration::from_secs(60);
 
 // The joining stage - node is waiting to be approved by the section.
 pub struct Joining {
     // EldersInfo of the section we are joining.
     elders_info: EldersInfo,
+    // PublicKey of the section we are joining.
+    section_key: bls::PublicKey,
     // Whether we are joining as infant or relocating.
     join_type: JoinType,
+    // Token for the join request timeout.
+    timer_token: u64,
+    // Message we can't handle until we are joined.
+    msg_backlog: Vec<QueuedMessage>,
 }
 
 impl Joining {
     pub fn new(
         core: &mut Core,
         elders_info: EldersInfo,
+        section_key: bls::PublicKey,
         relocate_payload: Option<RelocatePayload>,
+        msg_backlog: Vec<QueuedMessage>,
     ) -> Self {
         let join_type = match relocate_payload {
             Some(payload) => JoinType::Relocate(payload),
-            None => {
-                let timeout_token = core.timer.schedule(JOIN_TIMEOUT);
-                JoinType::First { timeout_token }
-            }
+            None => JoinType::First,
         };
+        let timer_token = core.timer.schedule(JOIN_TIMEOUT);
 
         let stage = Self {
             elders_info,
+            section_key,
             join_type,
+            timer_token,
+            msg_backlog,
         };
         stage.send_join_requests(core);
         stage
     }
 
-    // Returns whether the timeout was handled.
-    pub fn handle_timeout(&mut self, core: &mut Core, token: u64) -> bool {
-        let join_token = match self.join_type {
-            JoinType::First { timeout_token } => timeout_token,
-            JoinType::Relocate(_) => return false,
-        };
-
-        if join_token == token {
-            debug!("Timeout when trying to join a section.");
-
-            for addr in self
-                .elders_info
-                .elders
-                .values()
-                .map(|node| *node.peer_addr())
-            {
-                core.transport.disconnect(addr);
-            }
-
-            true
-        } else {
-            false
+    pub fn handle_timeout(&mut self, core: &mut Core, token: u64) {
+        if token == self.timer_token {
+            debug!("Timeout when trying to join a section");
+            // Try again
```

Comment:
The only reason why this would timeout is that the target section is going through heavy churn which should be only temporary. So the join should always eventually succeed, maybe after a couple of tries. Also the upper layer can always use its own timeout and handle it however it wants.

Reply:
ok